### PR TITLE
1296 add postcode level data

### PR DESCRIPTION
**Pull request by billli0**
### Issue #:
[#1296](https://github.com/beyondessential/tupaia-backlog/issues/1296)
### Changes:
-  Add 'postcode' in 'entity_type'
- Change covideau entity canonical types

### Manual step:
- Import postcode data in AU with [file](https://github.com/beyondessential/tupaia-backlog/issues/1296#issuecomment-857290079) 
- Add a new tracked entity type in DHIS: [DHIS tracked entity type management](https://dev-aggregation.tupaia.org/dhis-web-commons/security/login.action#/list/programSection/trackedEntityType)

### CODE REVIEW

File: packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    const point = { type: 'Point', coordinates: [long, lat] };
+    await insertObject(db, 'entity', entity);
+    await db.runSql(`
+        update entity 
+        SET point = ST_Force2D(ST_GeomFromGeoJSON('${JSON.stringify(point)}'))
+        where code = '${entity.code}';
+    `);
+    // Insert relationships into entity_relation table
+    await db.runSql(
+      `UPDATE entity_hierarchy SET canonical_types = '{country, district, sub_district, postcode}' WHERE name = '${HIERARCHY_CODE}';`,
```
Comment (rohan-bes) [inline packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js:66]: This works, right? Even though technically the hierarchy is branched such that:
```
     country
        |
      state
   |        |
postcode    LGA
```
Personally I do find it a touch ugly that we can support via such a definition, but it's a lot easier than defining entity-relations for the state<=>postcode and state<=>LGA mapping. Fine by me 👍

File: packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      country_code: 'AU',
+      metadata: {
+        dhis: {
+          isDataRegional: true,
+        },
+        openStreetMaps: {},
+      },
+    };
+    const point = { type: 'Point', coordinates: [long, lat] };
+    await insertObject(db, 'entity', entity);
```
Comment (rohan-bes) [inline packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js:58]: We aren't gonna run into any issues with meditrak syncing or anything due to manually inserting entities via a migration like this are we? @nick-beyondessential you had to work on something similar recently with the covid fiji entity import right? If I recall you imported the entities via the admin panel correct?

**Review by rohan-bes — COMMENTED** [review]
Just leaving this review as a comment for now as I have a recollection that we actually need to import the entities via the admin panel rather than via a migration. All the other logic to create the `postcode` entity type and update the entity hierarchy seems correct tho 👍

File: packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js (hunk 2 above)
Comment (nick-beyondessential) [reply packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js:58]: Yes best practise is to import via admin-panel for meditrak sideeffects, which may not apply in this particular case, mostly to do with adding facilies to tables etc, we can't be sure of catching all of them, so adding via admin is safer.

File: packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js (hunk 1 above)
Comment (nick-beyondessential) [reply packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js:66]: The canonical_types list is not order specific.

File: packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js (hunk 2 above)
Comment (billli0) [reply packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js:58]: Thanks @nick-beyondessential ! Could you share an example file or the right format for the import file with latitude and longitude?

Comment (nick-beyondessential) [reply packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js:58]: Something like this:

[importPostcodes.xlsx](https://github.com/beyondessential/tupaia/files/6613351/importPostcodes.xlsx)
https://beyond-essential.slab.com/posts/importing-entities-w8ewpm2b

Comment (nick-beyondessential) [reply packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js:58]: Seeing as you have already written the migration to input postcodes. Maybe one option would be to run locally then export entities and use that file as the input file.

Comment (billli0) [reply packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js:58]: Thanks @nick-beyondessential !

File: packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js (hunk 1 above)
Comment (rohan-bes) [reply packages/database/src/migrations/20210607053528-AddPostcodeLevelDataInAU-modifies-data.js:66]: That's true! I suppose I just find it confusing that it isn't 😝

File: packages/database/src/migrations/20210607052816-CreatePostCodeEntityType-modifies-schema.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  seed = seedLink;
+};
+
+exports.up = async function () {
+  const db = new TupaiaDatabase();
+  await db.executeSql(`
+    ALTER TYPE public.entity_type ADD VALUE 'postcode';
+  `);
+  return db.closeConnections();
+};
```
Comment (rohan-bes) [inline packages/database/src/migrations/20210607052816-CreatePostCodeEntityType-modifies-schema.js:25]:
```suggestion
exports.up = async function (db) {
  await db.executeSql(`
    ALTER TYPE public.entity_type ADD VALUE 'postcode';
  `);
};
```
Is there a reason we aren't doing this migration the normal way?

**Review by rohan-bes — APPROVED** [review]
Just one comment regarding the way we're running the CreatePostCodeEntityType migration, though I suppose it's fine?

Pre-approving anyway cos it's all looking good 👍

File: packages/database/src/migrations/20210607052816-CreatePostCodeEntityType-modifies-schema.js (hunk 3 above)
Comment (billli0) [reply packages/database/src/migrations/20210607052816-CreatePostCodeEntityType-modifies-schema.js:25]: tbh I don't know, but I think that is our traditional way to add an `entity.type` 😁

**Comment by billli0** [comment]
Will merge to epic branch `epic-flu-tracking-visuals` for testing

**Pull request merged**